syntax = "proto3";

package candig;

import "candig/variants.proto";
import "google/api/annotations.proto";

enum Genotype {
    HOMOZYGOUS_REF = 0;
    HETEROZYGOUS_ALT = 1;
    HOMOZYGOUS_ALT = 2;
    HEMIZYGOUS_REF = 3;
    HEMIZYGOUS_ALT = 4;
    OTHER = 5;
    NA = 6;
}

message GenotypeMatrix {
  uint32 nindividuals = 1;
  uint32 nvariants = 2;
  repeated Genotype genotypes = 3;
}

service GenotypesService {
  // Gets a list of `Genotypes` matching the search criteria.
  //
  // `POST /genotypes/search` must accept a JSON version of
  // `SearchGenotypesRequest` as the post body and will return a JSON version of
  // `SearchGenotpesResponse`.
  rpc SearchGenotypes(SearchGenotypesRequest)
    returns (SearchGenotypesResponse) {
      option (google.api.http) = {
        post: "/v0.8.0/genotypes/search"
        body: "*"
      };
    };
}

// ******************  /variants  *********************
// This request maps to the body of `POST /variants/search` as JSON.
message SearchGenotypesRequest {
  // The `VariantSet` to search.
  string variant_set_id = 1;

  // Only return variant calls which belong to call sets with these IDs.
  // If unspecified, return all variants and no variant call objects.
  repeated string call_set_ids = 2;

  // Required. Only return variants on this reference.
  string reference_name = 3;

  // Required. The beginning of the window (0-based, inclusive) for
  // which overlapping variants should be returned.
  // Genomic positions are non-negative integers less than reference length.
  // Requests spanning the join of circular genomes are represented as
  // two requests one on each side of the join (position 0).
  int64 start = 4;

  // Required. The end of the window (0-based, exclusive) for which overlapping
  // variants should be returned.
  int64 end = 5;

  // Specifies the maximum number of results to return in a single page.
  // If unspecified, a system default will be used.
  int32 page_size = 6;

  // The continuation token, which is used to page through large result sets.
  // To get the next page of results, set this parameter to the value of
  // `next_page_token` from the previous response.
  string page_token = 7;
}

message SearchGenotypesResponse {
  // The list of matching variants that we're returning genotypes for.
  repeated Variant variants = 1;

  // The list of callsets we're returning genotypes for
  repeated string call_set_ids = 2;

  GenotypeMatrix genotypes = 3;
  // The continuation token, which is used to page through large result sets.
  // Provide this value in a subsequent request to return the next page of
  // results. This field will be empty if there aren't any additional results.

  string next_page_token = 4;
}
